File: 704_binarysearch/binarysearch_recursive.swift
class Solution {
    func binarySearch(_ nums:[Int], _ target: Int,_ left: Int,_ right: Int) -> Int {
        if left > right {
            return -1
        }
        let pivot = left + (right - left) / 2
        if nums[pivot] == target {
            return pivot
        }
        if nums[pivot] > target {
            return binarySearch(nums, target, left, pivot - 1)
        }
        return binarySearch(nums, target, pivot + 1, right)
    }

    func search(_ nums: [Int], _ target: Int) -> Int {
        var pivot: Int
        var left = 0
        var right = nums.count - 1
        return binarySearch(nums, target, left, right)
    }
}

File: 704_binarysearch/binarysearch_recursive.ts
function search(nums: number[], target: number): number {
    let l = 0
    let r = nums.length - 1
    let mid = Math.floor((l + r) / 2)

    if (r % 2 !== 0) {
        mid += 1
    }

    if (nums[mid] === target) {
        return mid
    }
    if (nums.length === 1) {
        return -1
    }
    if (nums[mid] > target) {
        const result = search(nums.slice(0, mid), target)
        return result
    }
    const result = search(nums.slice(mid), target)
    return result === -1 ? -1 : result + mid
};
